Backport: add theme json class changes to edit style variations

tellthemachines opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: https://core.trac.wordpress.org/ticket/57583

 Backports changes from https://github.com/WordPress/gutenberg/pull/46343

These changes can't really be tested without the corresponding package updates, but I checked the Site Editor and Global Styles sidebar and at least nothing seems to be breaking 😅 


---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

tellthemachines committed:
Add theme json class changes to edit style variations

costdev commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -2124,6 +2152,48 @@ public function get_styles_for_block( $block_metadata ) {
+							// Compute declarations for the feature.
+							$new_feature_declarations = static::compute_style_properties( array( $feature_name => $style_variation_node[ $feature_name ] ), $settings, null, $this->theme_json );
+
+							// Merge new declarations with any that already exist for
+							// the feature selector. This may occur when multiple block
+							// support features use the same custom selector.
```
Multiline comments should be formatted as such:

```
/*
 * Line 1
 * Line 2
 * ...
 */
```
[Ref](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#5-2-multi-line-comments)
:1234: Applies elsewhere in the PR.

However, I note that this file contains many multiline comments that don't adhere to the standards.

In lieu of a change in this PR that creates an inconsistency in this file, I strongly suggest a follow up PR to bring this file up to the documentation coding standards, and that all future code being backported meets the standard.

hellofromtonya replied:
Resolved here https://github.com/WordPress/wordpress-develop/pull/3941/commits/0768aa8d5f8ee902bdc8d0ded9f326eb56ab6376.

costdev reviewed (commented):
Thanks for the PR @tellthemachines! I've just left a comment in this review regarding documentation standards for this file and future backports.

oandregal committed:
Add test for style variations

oandregal committed:
fixup

oandregal reviewed (approved):
I've tested this PR by doing the following:

- Go to TwentyTwentyThree's `theme.json` and under `styles.blocks.core/quote` add this: `"variations": { "plain": { "color": { "background": "hotpink"} } }`.
- Then go to the editor and add a post that contains a quote block.
- Set the quote to use the "plain" style variation and publish.
- The expected result is that the background color is hotpink.

I went ahead and prepared https://github.com/tellthemachines/wordpress-develop/pull/1 to add some unit test. It'd be great if it could be merged into this PR. Other than that and the comments brought up by @costdev , this is ready :ship:

tellthemachines commented:
Thanks for reviewing folks!

@costdev it's probably easiest to do a follow-up PR to fix all the comments once this one is in; I'm happy to do it.

@oandregal thanks for the test! I can't get it to pass locally but I'll merge it into this branch so we can see if it passes CI.

tellthemachines committed:
Merge pull request #1 from oandregal/backport/edit-style-variations

Unit test for Backport/edit style variations

tellthemachines committed:
Fix test output styles.

tellthemachines commented:
I removed the spaces between classnames from the output string as they're not meant to be there, but have no idea why the `Parse error: syntax error, unexpected ')' in /var/www/tests/phpunit/tests/theme/wpThemeJson.php on line 3620` - the tests are now passing for me locally 😕 

Edit: they now seem to be failing only on PHP versions 7.2 and below. Not sure why this would happen unless the actual issue is on a different line? That closing bracket it's complaining about seems to be in the right place.

tellthemachines committed:
Fix lint errors.

tellthemachines committed:
Fix spacing in test file.

oandregal commented on tests/phpunit/tests/theme/wpThemeJson.php:
```diff
@@ -3595,6 +3595,66 @@ public function test_get_styles_for_block_with_content_width() {
+						),
+					),
+				),
+			),
```
```suggestion
			)
```

oandregal replied:
Trailing commas in parameter list are allowed from 7.3 and above, so that may be the reason why the tests fail for anything below 7.2. I didn't catch that when I prepared the test locally, sorry about all the troubles in integrating it here!

hellofromtonya replied:
Let me take a look. I'll push changes to fix the tests.

hellofromtonya committed:
Remove trailing comma to resolve parsing error in < PHP 7.3.

Co-authored-by: André <583546+oandregal@users.noreply.github.com>

hellofromtonya committed:
Tests: split and use data provider

Splits the tests into 2 separate tests.

Adds a data provider for each.

Adds more datasets.

hellofromtonya commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -665,9 +665,17 @@ protected static function sanitize( $input, $valid_block_names, $valid_element_n
-			$schema_styles_blocks[ $block ]             = $styles_non_top_level;
-			$schema_styles_blocks[ $block ]['elements'] = $schema_styles_elements;
+			// Build the schema for each block style variation.
+			$style_variation_names    = isset( $input['styles']['blocks'][ $block ]['variations'] ) ? array_keys( $input['styles']['blocks'][ $block ]['variations'] ) : array();
```
What if the array structure exists, but the `'variations'` are not an array? What would happen?

https://3v4l.org/kHpHT

* >= PHP 8.0
```
Fatal error: Uncaught TypeError: array_keys(): Argument #1 ($array) must be of type array, string given
```
* > PHP 8.0:
```
Warning: array_keys() expects parameter 1 to be array, string given
```

To avoid these, also check if it is an array.

I'd also suggest switching to `! empty()` instead of `isset()`. Why? To avoid doing `array_keys()` on an empty array.

```suggestion
			$style_variation_names = array();
			if (
				! empty( $input['styles']['blocks'][ $block ]['variations'] ) &&
				is_array( $input['styles']['blocks'][ $block ]['variations'] )
			) {
				$style_variation_names = array_keys( $input['styles']['blocks'][ $block ]['variations'] );
			}
```

hellofromtonya replied:
Resolved here https://github.com/WordPress/wordpress-develop/pull/3941/commits/86ac232dd9f489a221770a9cdb35b0b4b07a1c7d.

hellofromtonya committed:
Guard for non-array variations

hellofromtonya commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -2039,12 +2063,23 @@ private static function get_block_nodes( $theme_json ) {
 			}
 
+			$variation_selectors = array();
+			if ( isset( $node['variations'] ) ) {
```
```suggestion
			if ( isset( $node['variations'] ) && is_array( $node['variations'] ) ) {
```

hellofromtonya commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -2124,6 +2159,48 @@ public function get_styles_for_block( $block_metadata ) {
+				// If the block has feature selectors, generate the declarations for them within the current style variation.
+				if ( ! empty( $block_metadata['features'] ) ) {
+					foreach ( $block_metadata['features'] as $feature_name => $feature_selector ) {
+						if ( ! empty( $style_variation_node[ $feature_name ] ) ) {
```
Unwrap the inner code by inverting this check. Why? Shifts the code to the left to make it more readable, as more levels of indentation can reduce readability.

hellofromtonya replied:
Resolved here https://github.com/WordPress/wordpress-develop/pull/3941/commits/994c20294df6ea20c57354b309fbcee48485eb9e.

hellofromtonya commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -2124,6 +2159,48 @@ public function get_styles_for_block( $block_metadata ) {
+							// Prepend the variation selector to the feature selector.
+							$split_feature_selectors    = explode( ',', $feature_selector );
+							$feature_selectors          = array_map(
+								function( $split_feature_selector ) use ( $style_variation_selector ) {
+									return trim( $style_variation_selector ) . trim( $split_feature_selector );
```
A few things here for micro-optimizations:

* Use `static function`. That way the callback is created once rather than on each iteration.

* Move the `trim( $style_variation_selector )` outside of the `foreach`. It does not change without the loop. By moving it outside of the loop, trimming happens once instead of on each iteration.

hellofromtonya replied:
Resolved here https://github.com/WordPress/wordpress-develop/pull/3941/commits/994c20294df6ea20c57354b309fbcee48485eb9e.

hellofromtonya committed:
Micro-optimizations in ::get_styles_for_block()

hellofromtonya committed:
Formatting for multiline comments

hellofromtonya reviewed (approved):
This PR LGTM 👍 It has test coverage, some micro-optimizations, and meets Core's coding standards.

Will wait for @tellthemachines to confirm the changes are okay before marking for `commit`.

hellofromtonya commented:
@tellthemachines Collectively there are some changes made in the PR for:

* Adding tests
* Multiline comment formatting
* Micro-optimizations
* Guarding to prevent PHP errors

The PR looks ready to me for `commit`. What do you think? Are the changes okay?

tellthemachines reviewed (commented):
Thanks for the improvements and detailed explanations @hellofromtonya ! The code looks good to me and tests are passing ✅

hellofromtonya commented:
Thanks @tellthemachines! Prepping commit now.

hellofromtonya commented:
Committed via https://core.trac.wordpress.org/changeset/55172.
